src/engine/GroupBy.cpp

```diff
@@ -388,15 +373,125 @@ ProtoResult GroupBy::computeResult([[maybe_unused]] bool requestLaziness) {
-      (std::array{inWidth, outWidth}), &GroupBy::doGroupBy, this,
-      subresult->idTable(), groupByCols, aggregates, &localVocab);
+  auto aggregateInfo = computeHashMapOptimizationMetadata(aggregates);
 
-  LOG(DEBUG) << "GroupBy result computation done." << std::endl;
-  return {std::move(idTable), resultSortedOn(), std::move(localVocab)};
+  if (!aggregateInfo.has_value()) {
+    // TODO<RobinTF> Should this throw an exception?
+    return {IdTable{getResultWidth(), getExecutionContext()->getAllocator()},
+            resultSortedOn(), std::move(localVocab)};
+  }
+
+  size_t size = getResultWidth() - groupByCols.size();
+  cppcoro::generator<IdTable> generator = CALL_FIXED_SIZE(
+      size, &GroupBy::computeResultLazily, this, std::move(subresult),
+      std::move(aggregateInfo).value().aggregateAliases_,
+      std::move(groupByCols));
+  return {std::move(generator), resultSortedOn(), std::move(localVocab)};
+}
+
+// _____________________________________________________________________________
+template <size_t NUM_AGGREGATED_COLS>
+cppcoro::generator<IdTable> GroupBy::computeResultLazily(
+    std::shared_ptr<const Result> subresult,
+    std::vector<HashMapAliasInformation> aggregateAliases,
+    std::vector<size_t> groupByCols) const {
+  AD_CONTRACT_CHECK(groupByCols.size() + NUM_AGGREGATED_COLS ==
+                    getResultWidth());
+
+  auto localVocab = subresult->getCopyOfLocalVocab();
+
+  LazyGroupBy<NUM_AGGREGATED_COLS> lazyGroupBy{
+      localVocab, std::move(aggregateAliases), groupByCols};
+
+  IdTable resultTable{getResultWidth(), getExecutionContext()->getAllocator()};
+
+  // TODO<RobinTF> Check if it is possible to not prolong the lifetime of this
+  // `IdTable` by seeing if the evaluation context really needs a proper
+  // `IdTable` to work.
+
+  // Extend the lifetime of the `IdTable` yielded by the generator a little bit
+  std::optional<IdTable> idTableWrapper = std::nullopt;
+  std::optional<sparqlExpression::EvaluationContext> evaluationContext =
+      std::nullopt;
+
+  for (IdTable& tmpIdTable : subresult->idTables()) {
+    if (tmpIdTable.empty()) {
+      continue;
+    }
+    idTableWrapper = std::move(tmpIdTable);
+    const IdTable& idTable = idTableWrapper.value();
+    checkCancellation();
+
+    lazyGroupBy.populateGroupBlock(idTable, 0);
+
+    evaluationContext.emplace(*getExecutionContext(),
+                              _subtree->getVariableColumns(), idTable,
+                              getExecutionContext()->getAllocator(), localVocab,
+                              cancellationHandle_, deadline_);
+
+    evaluationContext->_groupedVariables = ad_utility::HashSet<Variable>{
+        _groupByVariables.begin(), _groupByVariables.end()};
+    evaluationContext->_isPartOfGroupBy = true;
+    evaluationContext->_variableToColumnMapPreviousResults =
+        getInternallyVisibleVariableColumns();
+    evaluationContext->_previousResultsFromSameGroup.resize(getResultWidth());
```

Comment:
I feel that we now have this (complicated) setup of a GroupBy-specific evaluationContext multiple times
(for lazy/materialized/hashMap) with a lot of common code that depends only on the GroupBy class itself, 
Can you factor that out please?